design/app-to-app.md

```diff
@@ -0,0 +1,125 @@
+public offSharingMessage(messageType: string, callbacks: SharingCallbacks): bool {
+    // [... Deregister listener ...]
+    return true; // If successful
+}
+```
+
+## Implementation Suggestions
+
+1. Modify `context` to expose callback registration mechanism
+2. Adjust `contextInternal` and perhaps `multiPeerAdapter` to message the desire for each MessageType callback to each client (and corresponding remove messages). For callbacks registered for all sessions or all users, this my be sent after connect
+3. Implement the "Photo" message type. Send and receive encoded photos via the existing websocket used by multiPeerAdapter. Future message types (e.g., streaming video) may need to establish related side-channels.
+
+### Sample data flow for photo sharing
+1. On connect (or on subsequent sharing message registration), MRE app sends a `SharingRegistration` json payload consisting of
+    * `messageType` - `string` (as Described in [API Design](#API-Design))
+2. MRE Client sends a `SharingMessage` json payload, consiting of the following fields (as Described in [API Design](#API-Design))
+    * `messageId` - `string` (randomly generated guid)
+    * `messageType` - `string` 
+    * `photoType` - `string`
+    * `dataSize` - `number`
+    * `transmissionMethod` - `string`, presently set to the value "websocketbinary", but may be extended, for example, to support alternative trasnport for existing message types such as json messages on the primary channel (for smaller payloads)
+3. MRE app invokes `SharingCallbacks.isMessageDesired` callback if present to determine whether the message is desired. If a `SharingCallback` is registered for the given `messageType` but no `isMessageDesired` is present, proceed as if the message IS desired. If no `SharingCallback` is registered for the given `messageType`, proceed as if the message is NOT desired.
+4. MRE app sends a `SharingMessageResponse` json payload, consiting of the following fields:
+    * `result` - `boolean` true if the message is desired, false otherwise.
+5. If the message was desired, the MRE host should establish a [Binary Side Channel](#Binary-Side-Channel) and transmit the bytes of the image immediately following the `SharingMessageResponse` in binary format.
+6. MRE app invokes `SharingCallbacks.messageReceived` with a payload similar to that of step 2, with the additional binary data of the image on the payload.
+7. MRE app sends a `SharingMessageResponse` similar to step 4, indicating whether the message was successfully received by an appropriate handler
+
+Note that future sharing message types for content smaller than images may forgo steps 4 and 5 and elect to send their entire payload with the initial json sent in step 2.
+
+
+### Binary Side Channel
+Certain message types, such as images, are not suitable for transmission on the primary websocket. In these cases a websocket bound to an additional port (TODO: Pick port number) will accept connections to handle expected payloads.
```

Comment:
Additional ports for web sockets are not guaranteed for several hosting platforms, including Azure Web Apps.  They are sandboxed and only have ports 80 and 443 open for communication.  This seems like a less than ideal solution to this, as many of our own hosted apps are built on Azure Web App services.

Reply:
Thanks for this! Updated to restful api per your feedback!